### PR TITLE
Minor clarifications

### PR DESCRIPTION
**Pull request by toote**
After some comments in [issue 372 in the plugin linter](https://github.com/buildkite-plugins/buildkite-plugin-linter/issues/372) I believe that some clarifications are necessary on the documentation to write plugins with regards to the `builkite-plugin` suffix and vendored plugins.

### CODE REVIEW

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:52]: Is the _plugin ID_ defined or referenced anywhere else? Does it mean a lowercase version of the plugin name, with any whitespace replaced with dashes?

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:55]: Just a couple of nitpicks.

```suggestion
> Note that if you <a href="/docs/plugins/using#plugin-sources">reference a plugin</a> with a full URL ending in <code>.git</code> and that plugin's name does not end with `-buildkite-plugin`, variable names will include `_GIT` as part of the plugin name. For example, the value of the configuration <code>pattern</code> in <code>https://github.com/my-org/my-plugin.git#v1.0.0</code> will be available as <code>BUILDKITE_PLUGIN_MY_PLUGIN_GIT_PATTERN</code>.
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:54]: The rest of the page frames these ideas in terms of _properties_ rather than _variables_, so I think it would be good to stick to that. I think it could be more descriptive. Thinking:

```suggestion
>🚧 Accessing plugin properties with Git URLs
```

(_Git_ should also be capitalized)

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:100]: I don't think we need the repetition since it's clear from the plugin name.

```suggestion
### Plugin linter
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:102]: It would be good to be more specific here:

```suggestion
If your plugin has a Buildkite pipeline, you can add a step to lint it using the Plugin linter:
```

**Review by mbelton-buildkite — COMMENTED** [review]
Thanks for putting this together! Sorry it's taken so long to move forward. I've left some comments for how I think we can adjust it before merging.

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:54]: How about the follwing to avoid a potential confusion of Git URLs being the values of the plugin properties?

```suggestion
>🚧 Accessing properties on plugins referenced with Git URLs
```

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:100]: Not really... this is a buildkite plugin that runs the _linter for buildkite plugins_ so the plugin is not a repetition but an actual description that is necessary to differentiate between the linter itself and the plugin that runs the linter

Comment (toote) [reply pages/plugins/writing.md.erb:100]: Maybe a better wording would be:
```suggestion
### Buildkite Plugin to run the plugin linter
```

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:102]: Same here... we are talking about the buildkite plugin that runs the plugin linter (not the plugin linter itself). Yeah, I know...

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:112]: Indeed... the two sub headings are two ways to make it easier to run the docker command by hand: using a buildkite plugin or docker compose

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:52]: That is a good point... I don't think that there is a full documentation of how the ID is created. Environment variables are [created by the agent](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L249). Following the code, it would appear that - at this time at least - the logic is as follows:

* take [the plugin name](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L139):
  - from the plugin's definition, take whatever is after the last slash (`/`)
  - make it all lowercase
  - replace every whitespace in it with a single space
  - replace every character that is not a number or letter with a dash `-`
  - remove the `-buildkite-plugin(-git)` suffix
* convert it to a [variable name](https://github.com/buildkite/agent/blob/ec8675e062f4c9e83e772a827305b66331f1ba7a/agent/plugin/plugin.go#L207):
  - make it all uppercase
  - replace every whitespace in it with a single space
  - dashes and spaces are replaced with underscores
  - if there are contiguous underscores, they are replaced with a single one

So, basically, it is mostly what you said... except it is not the plugin name, but how it is referenced in the pipeline itself.

I used `plugin id` because that is how the linter defines it, but I don't think that it makes sense to use it.

How about the following?

```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but a variable-friendly (only letters, numbers and underscores) version of the repository or folder name in uppercase. In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb]: Looks good except for the link so I committed it myself.

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb]: I reworded this notice and the previous one to make the impact more clear

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:52]: Another alternative would be:

```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:30]: I think this is more of a note/tip than a warning.

```suggestion
> 📘 The `-buildkite-plugin` suffix
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:31]: I think this reads more clearly and breaks up the paragraph:

```suggestion
> We recommend using the `-buildkite-plugin` suffix in the repository name because:
<ul>
  <li>You can reference the plugin in pipelines using the <code>user/plugin-name</code> syntax rather than the full URL.</li>
  <li>It makes it easier for community members to find and use the plugin if you make it public.</li>
  <li>It communicates the purpose of the code.</li>
</ul>
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:54]: I think this is also more a note than a warning. We hope most people aren't running into this because they've followed the guidance above about including the suffix.

```suggestion
>📘 Accessing properties on plugins referenced with Git URLs
```

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:97]: I think we should replace all of `## Step 3: Validate the plugin.yml` with:

````
## Step 3: Validate the plugin.yml

The [Buildkite Plugin Linter](https://github.com/buildkite-plugins/buildkite-plugin-linter) is an app that helps ensure your plugin is up-to-date and has all the files required to list it in the plugins directory. The app is available as a Docker image you can run on the command line, with a dedicated plugin, or with Docker Compose. We recommend you start by running the linter on the command line, and then include the dedicated plugin in the pipeline for your plugin.

### Run on the command line

You can run the plugin linter with the following Docker command:

```shell
docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-linter --id a-github-user/file-counter
```

### Run with the dedicated plugin

If your plugin has a Buildkite pipeline, you can add a step to lint it using the corresponding plugin:

```yml
  - label: ":shell: Lint"
    plugins:
      plugin-linter#v3.0.0:
        id: a-gihub-user/file-counter
```
{: codeblock-file=".buildkite/pipeline.yml"}

### Run with Docker Compose

If you want to run the linter using Docker Compose, you can add the following to a `docker-compose.yml` file:

```yml
services:
  lint:
    image: buildkite/plugin-linter
    command: ['--id', 'a-github-user/file-counter']
    volumes:
      - ".:/plugin:ro"
```
{: codeblock-file="docker-compose.yml"} 

You can then run the tests using the following command:

```shell
docker-compose run --rm lint
```
````

What do you think?

**Review by mbelton-buildkite — COMMENTED** [review]
I tried to push a commit but was told I didn't have write access. These are the changes I think we should make.

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb]: I didn't quite like the mix of HTML... but could not find a work around it :(

File: pages/plugins/writing.md.erb
Comment (toote) [reply pages/plugins/writing.md.erb:104]: Used as-is, except for the main heading that I changed to `Validate the plugin` (as the linter checks more things than just the `plugin.yml` file)

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:57]: One last nitpick: Use a serial comma.
```suggestion
Configuration properties are available to the hook script as environment variables with the naming pattern `BUILDKITE_PLUGIN_<PLUGIN_NAME>_<CONFIGURATION_PROPERTY>` where `<PLUGIN_NAME>` is not the name defined in `plugin.yml` but the repository or folder name compatible with Bash environment variables (in uppercase and only letters, numbers, and underscores). In this case, the configured value of `pattern` will be available as `BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN`.
```

**Review by mbelton-buildkite — APPROVED** [review]
Looks good to me! Great work on this 🙌🏻

I'm just running the tests and then I'll merge for you (I don't think you have all the permissions to kick off a deploy).

File: pages/plugins/writing.md.erb
Comment (mbelton-buildkite) [inline pages/plugins/writing.md.erb:279]: Fix broken link
```suggestion
Vendored plugins run after non-vendored plugins and don't have access to all the same hooks. See [the documentation about job lifecycle hooks](/docs/agent/v3/hooks#job-lifecycle-hooks) to learn more.
```

**Pull request merged**
